src/tests/p11test/p11test_case_common.c

```diff
@@ -577,49 +551,50 @@ int callback_public_keys(test_certs_t *objects,
 #if OPENSSL_VERSION_NUMBER < 0x30000000L
 			EC_KEY *ec = EC_KEY_new_by_curve_name(nid);
 			EC_KEY_set_public_key(ec, ecpoint);
-			EC_POINT_free(ecpoint);
 			EC_KEY_set_group(ec, ecgroup);
-			EC_GROUP_free(ecgroup);
 			EVP_PKEY_set1_EC_KEY(o->key, ec);
 			EC_KEY_free(ec);
 #else
 			ctx = EVP_PKEY_CTX_new_from_name(0, "EC", 0);
-			char curve_name[80]; size_t curve_name_len = 0;
-			unsigned char pubkey[80]; size_t pubkey_len = 0;
-			if (EVP_PKEY_get_group_name(o->key, curve_name, sizeof(curve_name), &curve_name_len)||
-				EVP_PKEY_get_octet_string_param(o->key, OSSL_PKEY_PARAM_PUB_KEY, pubkey, sizeof(pubkey), &pubkey_len) != 1) {
-				debug_print(" [WARN %s ] Can not get params from EVP_PKEY", o->id_str);
-				EC_GROUP_free(ecgroup);
-				EC_POINT_free(ecpoint);
-				return -1;
-			}
 
+			const char *curve_name = OBJ_nid2sn(nid);
 			if (!(bld = OSSL_PARAM_BLD_new()) ||
-				OSSL_PARAM_BLD_push_utf8_string(bld, "group", curve_name, curve_name_len) != 1 ||
-				OSSL_PARAM_BLD_push_octet_string(bld, "pub", pubkey, pubkey_len) != 1 ||
+				OSSL_PARAM_BLD_push_utf8_string(bld, "group", curve_name, strlen(curve_name)) != 1 ||
+				OSSL_PARAM_BLD_push_octet_string(bld, "pub", pub, pub_len) != 1 ||
 				!(params = OSSL_PARAM_BLD_to_param(bld))) {
-				debug_print(" [WARN %s ] Can not set params from EVP_PKEY", o->id_str);
-				EC_GROUP_free(ecgroup);
-				EC_POINT_free(ecpoint);
-				OSSL_PARAM_BLD_free(bld);
-				EVP_PKEY_CTX_free(ctx);
-				return -1;
+				debug_print(" [WARN %s ] Cannot set params from EVP_PKEY", o->id_str);
+				goto ec_error;
 			}
-			OSSL_PARAM_BLD_free(bld);
 
 			if (ctx == NULL || params == NULL ||
 				EVP_PKEY_fromdata_init(ctx) != 1 ||
 				EVP_PKEY_fromdata(ctx, &o->key, EVP_PKEY_PUBLIC_KEY, params) != 1) {
-				debug_print(" [WARN %s ] Can not set params for EVP_PKEY", o->id_str);
-				EC_GROUP_free(ecgroup);
-				EC_POINT_free(ecpoint);
-				EVP_PKEY_CTX_free(ctx);
-				return -1;
+				debug_print(" [WARN %s ] Cannot set params for EVP_PKEY", o->id_str);
+				goto ec_error;
 			}
-			EVP_PKEY_CTX_free(ctx);
-			OSSL_PARAM_free(params);
 #endif
 		}
+
+		int ec_error = 0;
+		goto ec_out;
+
+	ec_error:
+		ec_error = 1;
+
+	ec_out:
+		if (pub_asn1) ASN1_STRING_free(pub_asn1);
+		if (ecgroup) EC_GROUP_free(ecgroup);
+		if (ecpoint) EC_POINT_free(ecpoint);
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+		if (ctx) EVP_PKEY_CTX_free(ctx);
+		if (bld) OSSL_PARAM_BLD_free(bld);
+		if (params) OSSL_PARAM_free(params);
+		if (cert_group) EC_GROUP_free(cert_group);
+		if (cert_point) EC_POINT_free(cert_point);
```

Comment:
The `if` should not be needed if the pointer is initially initialized to NULL as all of these function should correctly handle the NULL arguments (and when it was already freed, the `if` condition will not save you).

Reply:
Unnecessary `if`s removed.